File: clients/cpp/debugging/Camera.cpp
#include "Camera.hpp"

namespace debugging {

Camera::Camera(model::Vec2 center, double rotation, double attack, double fov) : center(center), rotation(rotation), attack(attack), fov(fov) { }

// Read Camera from input stream
Camera Camera::readFrom(InputStream& stream) {
    model::Vec2 center = model::Vec2::readFrom(stream);
    double rotation = stream.readDouble();
    double attack = stream.readDouble();
    double fov = stream.readDouble();
    return Camera(center, rotation, attack, fov);
}

// Write Camera to output stream
void Camera::writeTo(OutputStream& stream) const {
    center.writeTo(stream);
    stream.write(rotation);
    stream.write(attack);
    stream.write(fov);
}

// Get string representation of Camera
std::string Camera::toString() const {
    std::stringstream ss;
    ss << "Camera { ";
    ss << "center: ";
    ss << center.toString();
    ss << ", ";
    ss << "rotation: ";
    ss << rotation;
    ss << ", ";
    ss << "attack: ";
    ss << attack;
    ss << ", ";
    ss << "fov: ";
    ss << fov;
    ss << " }";
    return ss.str();
}

}

File: clients/php/Debugging/Camera.php
<?php

namespace Debugging {
    require_once 'Model/Vec2.php';
    require_once 'Stream.php';

    /**
     * Camera state
     */
    class Camera
    {
        /**
         * Center
         */
        public \Model\Vec2 $center;
        /**
         * Rotation
         */
        public float $rotation;
        /**
         * Attack angle
         */
        public float $attack;
        /**
         * Vertical field of view
         */
        public float $fov;
    
        function __construct(\Model\Vec2 $center, float $rotation, float $attack, float $fov)
        {
            $this->center = $center;
            $this->rotation = $rotation;
            $this->attack = $attack;
            $this->fov = $fov;
        }
    
        /**
         * Read Camera from input stream
         */
        public static function readFrom(\InputStream $stream): Camera
        {
            $center = \Model\Vec2::readFrom($stream);
            $rotation = $stream->readDouble();
            $attack = $stream->readDouble();
            $fov = $stream->readDouble();
            return new Camera($center, $rotation, $attack, $fov);
        }
        
        /**
         * Write Camera to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $this->center->writeTo($stream);
            $stream->writeDouble($this->rotation);
            $stream->writeDouble($this->attack);
            $stream->writeDouble($this->fov);
        }
    }
}